dockers/docker-fpm-frr/bgpcfgd

```diff
@@ -1,64 +1,117 @@
 #!/usr/bin/env python
 
 import sys
+import copy
+import Queue
 import redis
 import subprocess
 import syslog
-from swsssdk import ConfigDBConnector
+import os
+from swsscommon import swsscommon
 
-class BGPConfigDaemon:
 
-    def __init__(self):
-        self.config_db = ConfigDBConnector()
-        self.config_db.connect()
-        self.bgp_asn = self.config_db.get_entry('DEVICE_METADATA', 'localhost')['bgp_asn']
-        self.bgp_neighbor = self.config_db.get_table('BGP_NEIGHBOR')
+def run_command(command):
+    syslog.syslog(syslog.LOG_DEBUG, "[bgp cfgd] execute command {}.".format(command))
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
+    stdout = p.communicate()[0]
+    p.wait()
+    if p.returncode != 0:
+        syslog.syslog(syslog.LOG_ERR, '[bgp cfgd] command execution returned {}. Command: "{}", stdout: "{}"'.format(p.returncode, command, stdout))
+
+
+class BGPConfigManager(object):
+    def __init__(self, daemon):
+        self.daemon = daemon
+        self.bgp_asn = None
+        self.bgp_message = Queue.Queue(0)
+        daemon.add_manager(swsscommon.CONFIG_DB, swsscommon.CFG_DEVICE_METADATA_TABLE_NAME, self.__metadata_handler)
+        daemon.add_manager(swsscommon.CONFIG_DB, swsscommon.CFG_BGP_NEIGHBOR_TABLE_NAME, self.__bgp_handler)
+
+    def __metadata_handler(self, key, op, data):
+        if key != "localhost" \
+            or "bgp_asn" not in data \
+                or self.bgp_asn == data["bgp_asn"]:
+            return
+        syslog.syslog(syslog.LOG_INFO, '[bgp cfgd] ASN changed to {} from {}, restart BGP...'.format(data['bgp_asn'], self.bgp_asn))
+        run_command("supervisorctl restart start.sh")
+        run_command("service frr restart")
+        self.bgp_asn = data["bgp_asn"]
+    
+    def __bgp_handler(self, key, op, data):
+        self.bgp_message.put((key, op, data))
+        # If ASN was setted
+        if self.bgp_asn == None:
```

Comment:
I don't think the code matches the comment.

Reply:
I have polished the comment, could you help me to check whether it's clear?